feat: Chroma - allow remote HTTP connection

alperkaya opened the pull request:
### Related Issues

- fixes #1071
### Proposed Changes:

A remote http client is added
 <!--- In case of a feature: Describe what did you add and how it works -->

### How did you test it?

<!-- unit tests, integration tests, manual verification, instructions for manual tests -->
Manual test

### Notes for the reviewer

<!-- E.g. point out section where the reviewer  -->

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

anakin87 commented on integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:
```diff
@@ -79,9 +83,21 @@ def __init__(
         # Create the client instance
-        if persist_path is None:
-            self._chroma_client = chromadb.Client()
+        # Ensure only one of persist_path or host is provided
+        if persist_path and host:
+            error_message = "You cannot provide both `persist_path` and `host`. Please choose one."
```
```suggestion
        if persist_path and (host or port is not None):
            error_message = "You must specify `persist_path` for local persistent storage or, alternatively, `host` and `port` for remote HTTP client connection. You cannot specify both options."
```

let's check also `port`.
probably the `error_message` should be split over multiple lines.

anakin87 commented on integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:
```diff
@@ -79,9 +83,21 @@ def __init__(
+        if persist_path and host:
+            error_message = "You cannot provide both `persist_path` and `host`. Please choose one."
+            raise ValueError(error_message)
+        if host:
+            # Remote connection via HTTP client
+            self._chroma_client = chromadb.HttpClient(
+                host=host,
+                port=port or 8000,
+            )
```
```suggestion
        if host and port is not None:
            # Remote connection via HTTP client
            self._chroma_client = chromadb.HttpClient(
                host=host,
                port=port,
            )
```

Let's be explicit: in case users want to connect to a remote Chroma instance, they should specify both `host` and `port`.

anakin87 commented on integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:
```diff
@@ -79,9 +83,21 @@ def __init__(
+            )
+        elif persist_path is None:
+            # In-memory storage
+            self._chroma_client = chromadb.Client()  # In-memory client
```
```suggestion
            self._chroma_client = chromadb.Client()
```
nit: redundant comment

anakin87 reviewed (changes requested):
Thanks for the contribution!

I found some small opportunities for improvement.

Also, let's **test the new behavior**.
This can be done as follows:
- add a new step in the [Chroma testing workflow](https://github.com/deepset-ai/haystack-core-integrations/blob/main/.github/workflows/chroma.yml) before `Run tests` to run the DB server as described in [Chroma docs](https://docs.trychroma.com/guides#running-chroma-in-client-server-mode).
Something like this should work: `hatch run chroma run`

- add a new test in https://github.com/deepset-ai/haystack-core-integrations/blob/main/integrations/chroma/tests/test_document_store.py to test the different init options (one of them is the newly introduced remote Chroma).

In case you need help adding the test, LMK and I can help/do it.

Last note: the introduction of this connection option conflicts with #746.
So after this PR is merged, I can take care of adapting the Document Store to defer the DB connection.

alperkaya commented:
> Thanks for the contribution!
> 
> I found some small opportunities for improvement.
> 
> Also, let's **test the new behavior**. This can be done as follows:
> 
>     * add a new step in the [Chroma testing workflow](https://github.com/deepset-ai/haystack-core-integrations/blob/main/.github/workflows/chroma.yml) before `Run tests` to run the DB server as described in [Chroma docs](https://docs.trychroma.com/guides#running-chroma-in-client-server-mode).
>       Something like this should work: `hatch run chroma run`
> 
>     * add a new test in https://github.com/deepset-ai/haystack-core-integrations/blob/main/integrations/chroma/tests/test_document_store.py to test the different init options (one of them is the newly introduced remote Chroma).
> 
> 
> In case you need help adding the test, LMK and I can help/do it.
> 
> Last note: the introduction of this connection option conflicts with #746. So after this PR is merged, I can take care of adapting the Document Store to defer the DB connection.

Hi,

Thanks for feedback!

I made a slight change in __init__() to make it more testable, added some test to check different init configs, and updated to to_dict() to support http client as well.

anakin87 commented:
I will have a look later...

In the meantime, to simplify things, I think we can test the HTTP client on Linux only (and skip the corresponding test on Windows).

anakin87 reviewed (approved):
I have taken the liberty of making some simplifications.
Thanks for the contribution!

I will merge this PR but I won't immediately release a new version of this package.

As I said before, I want to take care to make this Document Store consistent with others, by deferring the DB connection (see #746). I will do this in another PR.
